### PR TITLE
feat: Add deep link and launcher shortcut support for conversations

### LINKED ISSUE #847: Open Talk links in App
I have installed latest Nextcloud Talk app (F-Droid) and opened a talk link (`https://talk.example.com/call/pasvtaxg`).

Expected behavior: Talk Android app opens

Actual behavior: Browser opened.

Similar to https://github.com/stefan-niedermann/nextcloud-deck/issues/498

cc @jancborchardt @nextcloud/designers for a great UX :wink:

### PR DESCRIPTION
**Pull request by angrymuesli**
- Add custom URI scheme (nctalk://conversation/{token}) for opening conversations from external launchers like KISS
- Add HTTPS deep link support for /call/{token} URLs (fixes #847)
- Add dynamic shortcuts for favorite/recent conversations
- Add "Add to home screen" menu option in conversation long-press dialog
- New DeepLinkHandler utility for parsing deep link URIs
- New ShortcutManagerHelper utility for managing conversation shortcuts

### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
B | A

### 🚧 TODO

- [ ] ...

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included or not needed
- [ ] 🔖 Capability is checked or not needed 
- [ ] 🔙 Backport requests are created or not needed: `/backport to stable-xx.x`
- [ ] 📅 Milestone is set
- [ ] 🌸 PR title is meaningful (if it should be in the changelog: is it meaningful to users?)

### CODE REVIEW

**Comment by sowjanyakch** [comment]
Thank you for the PR @angrymuesli 
We will review your PR soon.

**Comment by sowjanyakch** [comment]
@angrymuesli - I tested this PR, shortcut support for conversations work well.  Opening Talk links from different domains and launching conversations via the external KISS launcher works only on Android 12 and earlier devices but not on Android 12+ devices.  Deep links only work if the domain is registered via assetlinks.json. Talk is a self-hosted and can run on many different domains, it is not feasible for us to register and verify all possible domains. As a result, automatically opening Talk links from different domains in the app on Android 12+ devices is practically not implementable.

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -122,6 +122,33 @@
...
+            </intent-filter>
+
+            <intent-filter android:autoVerify="true">
+                <action android:name="android.intent.action.VIEW" />
+                <category android:name="android.intent.category.DEFAULT" />
+                <category android:name="android.intent.category.BROWSABLE" />
+                <data android:scheme="https" />
+                <data android:host="*" />
+                <data android:pathPrefix="/index.php/call/" />
+            </intent-filter>
```
Comment (sowjanyakch) [inline app/src/main/AndroidManifest.xml:137]: The intent filter for opening talk conversation links can be removed because it does not work for Android 12+ devices.

**Comment by sowjanyakch** [comment]
To resolve linter issues, please run _./gradlew ktlintFormat_

There are also some static code analysis issues, see https://app.codacy.com/gh/nextcloud/talk-android/pull-requests/5664/issues. Please feel free to fix them, or let us know if you want us to handle them.

**Comment by angrymuesli** [comment]
> @angrymuesli - I tested this PR, shortcut support for conversations work well. Opening Talk links from different domains and launching conversations via the external KISS launcher works only on Android 12 and earlier devices but not on Android 12+ devices. Deep links only work if the domain is registered via assetlinks.json. Talk is a self-hosted and can run on many different domains, it is not feasible for us to register and verify all possible domains. As a result, automatically opening Talk links from different domains in the app on Android 12+ devices is practically not implementable.

Hi, and thanks for the feedback! Not all of that is true. I installed a debug build based on the latest changes on my Pixel 6a running GrapheneOS (Android 16), and I could easily access Talk conversations in KISS by typing a few letters of the conversation name. Did you mean something else?

**Comment by angrymuesli** [comment]
> To resolve linter issues, please run _./gradlew ktlintFormat_
> 
> There are also some static code analysis issues, see https://app.codacy.com/gh/nextcloud/talk-android/pull-requests/5664/issues. Please feel free to fix them, or let us know if you want us to handle them.

Sure, I'll run the linter and address the static code analysis issues. Thanks for pointing these out.

**Comment by sowjanyakch** [comment]
> Hi, and thanks for the feedback! Not all of that is true. I installed a debug build based on the latest changes on my Pixel 6a running GrapheneOS (Android 16), and I could easily access Talk conversations in KISS by typing a few letters of the conversation name. Did you mean something else?

Talk conversations could be accessed on KISS launcher, however i noticed a issue - When I try to access a conversation, sometimes a different conversation opens, sometimes the requested conversation opens and sometimes it just opens the conversations list.  Please see the below video and I used single account for testing purpose. 

https://github.com/user-attachments/assets/357cc8ad-b586-4e6a-a82f-4c0d2cff8b7e

Additionally, the other thing, that does not work is opening different domain talk links on the talk app, it redirects to web browser instead of app for Android 12+ devices.

**Comment by mahibi** [comment]
Thank you @angrymuesli :+1: 

@sowjanyakch It's important to differ between https deep links (App Links), custom URI schemes and dynamic shortcuts. All solutions are implemented in this PR.

https deep links come with the domain ownership verification problem (assetlinks.json) for newer android versions.
This is not suitable for us because each instance has a different domain.
It would only work if every domain owner would build their own version of the nextcloud talk app and take care of the verification.
I'm not sure if we're doing ourselves any favors by starting to support https deep links for older Android versions now. Users who get used to this will complain as soon as their Android version is updated and the links no longer work.

This is why we must go for custom URI schemes.
Like angrymuesli suggests 
`nctalk://`

However we have to be careful with implementing as it should be common for all platforms.
There is already the issue created by @ShGKme :
- https://github.com/nextcloud/spreed/issues/16354

There the suggestion is 
`nextcloudtalk:// `
as protocol name.
And it's not only the protocol name that we have to define but also the complete schema.
This is not done yet or at least there has to be the final decision from @nickvergessen .
This is why i will decline the pull request for now.
But don't get me wrong: The pull request is great and thanks for contibuting @angrymuesli 
It just needs alignment with the other platforms regarding the schema.
We will make sure the schema is discussed and a final decision is made. Most probably the pull request will have to be modified because of this.

Todos:
- [x] decision if implementing https deep links makes sense at all as they will only work for android 12 and lower.
- [x] define custom URI scheme

**Comment by mahibi** [comment]
> > @angrymuesli - I tested this PR, shortcut support for conversations work well. Opening Talk links from different domains and launching conversations via the external KISS launcher works only on Android 12 and earlier devices but not on Android 12+ devices. Deep links only work if the domain is registered via assetlinks.json. Talk is a self-hosted and can run on many different domains, it is not feasible for us to register and verify all possible domains. As a result, automatically opening Talk links from different domains in the app on Android 12+ devices is practically not implementable.
> 
> Hi, and thanks for the feedback! Not all of that is true. I installed a debug build based on the latest changes on my Pixel 6a running GrapheneOS (Android 16), and I could easily access Talk conversations in KISS by typing a few letters of the conversation name. Did you mean something else?

Dynamic shortcuts work by intents as seen in 
https://github.com/nextcloud/talk-android/pull/5664/changes#diff-26c7d7916feb623193bb39f5986bf3217e4eab7a7b559ef078810f5e4a1a1a55R61
so it's neither a https deep link nor a custom URI scheme.
That's why it works.

> When I try to access a conversation, sometimes a different conversation opens, sometimes the requested conversation opens and sometimes it just opens the conversations list.

I did not try with KISS launcher (so could also be a bug of the launcher..). For the native android launcher it worked for me.

**Review by mahibi — CHANGES_REQUESTED** [review]
see comment from https://github.com/nextcloud/talk-android/pull/5664#issuecomment-3759835609

**Comment by mahibi** [comment]
HI @angrymuesli 

we decided that the schema that is described in 
https://github.com/nextcloud/spreed/issues/16354
will be taken as is. 

Also, i suggest that implementing https deep links should not be done as it's already considered as legacy.

Could you implement these changes? Just let us know if there are any questions!

**Comment by angrymuesli** [comment]
> HI @angrymuesli
> 
> we decided that the schema that is described in [nextcloud/spreed#16354](https://github.com/nextcloud/spreed/issues/16354) will be taken as is.
> 
> Also, i suggest that implementing https deep links should not be done as it's already considered as legacy.
> 
> Could you implement these changes? Just let us know if there are any questions!

Hi @mahibi,
Thank you for all the comments!

I'm glad we've reached a consensus and are moving forward. I'll carefully read through them and update the pull request accordingly.

**Comment by mahibi** [comment]
thank you for continuing @angrymuesli :)
@sowjanyakch can you take care of this PR and follow/review it until it's merged?

**Comment by sowjanyakch** [comment]
@angrymuesli - please re-request a review after the PR is ready, I will review it.

**Comment by sowjanyakch** [comment]
Hello @angrymuesli,

Just checking in on this PR. Please let us know if you plan to continue, or if you’d like us to take over.
Happy to help if you have any questions.

**Comment by angrymuesli** [comment]
Hi @sowjanyakch, yes I'm continuing! Just force-pushed with:
- Rebased on latest master
- Fixed all Codacy static analysis issues (ReturnCount)
- Added DCO sign-off
- ktlint and detekt pass locally

The PR implements the `nextcloudtalk://` URI scheme as decided in spreed#16354, with no HTTPS deep links. Ready for review when you have time!

**Review by sowjanyakch — CHANGES_REQUESTED** [review]
1.  Please align the menu icon and title with other options.

Image

2. Deleting a conversation should also remove shortcut or make the shortcut disabled and show a warning shortcut is unavailable/conversation no longer available when clicked.

3. When multiple accounts A, B are present, create a shortcut for a conversation belonging to account A and then log out of account A. If the shortcut is opened, the app correctly opens the conversation list for account B. Now log in to account A and tap the same shortcut again, the app now switches to account A but displays the conversation list of account B instead of navigating to the intended conversation.

Image

**Comment by angrymuesli** [comment]
Hi @sowjanyakch, thank you for the detailed review! I've addressed all three issues:

1. **Menu alignment** — Fixed padding for "Add to home screen" to use `standard_dialog_padding` consistently with other items.

2. **Shortcut cleanup on delete** — Deleting (or leaving) a conversation now disables its shortcuts. Pinned shortcuts will show "This conversation no longer exists" when tapped.

3. **Multi-account shortcut bug** — Shortcuts now use `nextcloudtalk://` deep link URIs instead of bundle extras. This routes them through the deep link handler which properly resolves the correct user account before navigating to the conversation.

Also rebased on latest master. All checks pass locally (ktlint, detekt, unit tests).

File: app/src/main/java/com/nextcloud/talk/activities/MainActivity.kt [hunk 2]
```diff
@@ -253,34 +259,146 @@ class MainActivity :
...
+                            ShortcutManagerHelper.reportShortcutUsed(
+                                context,
+                                deepLinkResult.roomToken,
+                                userId
+                            )
+                        }
+
+                        if (isFinishing || isDestroyed) return@subscribe
+
+                        val chatIntent = Intent(context, ChatActivity::class.java)
```
Comment (sowjanyakch) [inline app/src/main/java/com/nextcloud/talk/activities/MainActivity.kt:352]: Login with two accounts Account A, B on talk. Set Account A as active user and create a shortcut for a conversation P. Switch active user to account B and then click on the shortcut - From the current implementation, switches the active user back to Account A and navigates to the intended chat. When you press back button, the conversations from account B are still loaded (conversation from target user are not shown). This creates a lot of bugs.
 
We could open the conversations list for the target user first and then try to navigate to the chat - then back press opens the conversation list for the target user.

**Comment by sowjanyakch** [comment]
Conversation deletion (or leaving) through ConversationListBottomDialog already disables the corresponding shortcut. However, conversations can also be deleted from ConversationInfoActivity, so shortcut disabling should be handled there as well.

**Comment by angrymuesli** [comment]
Hi @sowjanyakch, good catch on both issues! Fixed in the latest push:

1. **ConversationInfoActivity** — Shortcut disabling now also covers delete and leave from the conversation info screen, not just the bottom dialog.

2. **Back-stack** — Deep link navigation now uses `startActivities()` to place `ConversationsListActivity` underneath `ChatActivity`. Pressing back now correctly shows the target user's conversation list.

**Comment by sowjanyakch** [comment]
Thank you @angrymuesli for fixing the issues. Sorry for the late reply as I was on vacation for last two weeks. Could you fix the merge conflicts and rebase on the latest master so I can test this PR?

**Comment by angrymuesli** [comment]
Hi @sowjanyakch, welcome back! Rebased on latest master — ready for testing.

**Comment by sowjanyakch** [comment]
> 2. **Back-stack** — Deep link navigation now uses `startActivities()` to place `ConversationsListActivity` underneath `ChatActivity`. Pressing back now correctly shows the target user's conversation list.

The current user is not properly set to the target user before navigating to the conversation list, so the issue still persists.

**Comment by angrymuesli** [comment]
Hi @sowjanyakch, pushed with the fixes:

- **Back-stack fix**: ConversationsListActivity now reads the target user directly from the intent via `KEY_INTERNAL_USER_ID`, bypassing the stale provider cache. Pressing back from a shortcut-opened chat now shows the correct user's conversation list.
- **Compose migration**: Adapted "Add to home screen" to the new ModalBottomSheet architecture (`ConversationOpsAction` + `ConversationOperationsSheet`).
- **Shortcut cleanup**: All delete/leave paths now covered (ConversationsListActivity + ConversationInfoActivity).

Rebased on latest master, all checks pass locally. Looking forward to your feedback whenever you get a chance!

**Comment by mahibi** [comment]
Thank you for keeping it up @angrymuesli we will review again soon

**Comment by sowjanyakch** [comment]
Tested and works as expected.

Note to myself: The existing bug with the account switching also impacts this PR and it needs to be fixed.

**Comment by angrymuesli** [comment]
> Tested and works as expected.
> 
> Note to myself: The existing bug with the account switching also impacts this PR and it needs to be fixed.

Could you share the details of this account switcher bug, maybe that could be easily fixed within this PR?

**Comment by sowjanyakch** [comment]
> Could you share the details of this account switcher bug, maybe that could be easily fixed within this PR?

When using multi account, sometimes chats from another account are sometimes displayed in the current account. This is an existing bug and something we need to investigate.

**Comment by angrymuesli** [comment]
Thanks @sowjanyakch — looks like this is already tracked in #4039. Happy to help investigate in a separate PR once the root cause is identified.

I've rebased on latest master to resolve the new conflicts. Since you've confirmed it works as expected, could you formally approve when you have a moment? There are some older review states from earlier feedback rounds that GitHub is still showing as blocking — your approval would let the team move forward with the merge. Thanks!

**Comment by sowjanyakch** [comment]
@angrymuesli - I made some changes to the PR covering some edge cases like username will be xyz or xyz@cloud.com. This PR handled only first case due to wrong parsing. I will deep test again with multiple accounts from different servers and after that it's good to merge.

**Comment by sowjanyakch** [comment]
Notes : This PR works as expected but there is a problem with the account switching. Currently account switching does not work properly and it is observed multiple times on the latest master. This PR can be merged but can only be included in release once we fix the problem with the account switching.

**Review by sowjanyakch — APPROVED** [review]

**Comment by mahibi** [comment]
i took over in https://github.com/nextcloud/talk-android/pull/6207 to rebase on master, fix conflicts and also fixed a bug with opening multiple app instances when clicking a deep link.
Hope that's okay for you @angrymuesli 
Thanks again to this great contribution :+1:
